pairinteraction/SystemTwo.cpp

```diff
@@ -225,170 +236,182 @@ void SystemTwo::initializeBasis() {
     int parityM = 0;
 
     size_t col_new = 0;
-    for (size_t col_1 = 0; col_1 < system1.getNumBasisvectors(); ++col_1) {
-        for (size_t col_2 = 0; col_2 < system2.getNumBasisvectors(); ++col_2) {
-
-            // In case of artificial states, some symmetries won't work
-            auto sym_inversion_local = sym_inversion;
-            auto sym_reflection_local = sym_reflection;
-            auto sym_rotation_local = sym_rotation;
-            if (artificial1[col_1] || artificial2[col_2]) {
-                if (sym_inversion_local != NA || sym_reflection_local != NA ||
-                    sym_rotation_local.count(ARB) == 0) {
-                    std::cerr
-                        << "WARNING: Only permutation symmetry can be applied to artificial states."
-                        << std::endl;
-                }
-                sym_inversion_local = NA;
-                sym_reflection_local = NA;
-                sym_rotation_local = std::set<int>({ARB});
-            }
 
-            // In case of inversion or permutation symmetry: skip half of the basis vector pairs
-            if ((sym_inversion_local == EVEN && col_1 <= col_2) || // gerade
-                (sym_inversion_local == ODD && col_1 < col_2)) {   // ungerade
-                continue;
-            }
-            if ((sym_permutation == EVEN && col_1 <= col_2) || // sym
-                (sym_permutation == ODD && col_1 < col_2)) {   // asym
-                continue;
+    // Generate vector containing pairs of indices of one atom basisvectors
+    std::vector<std::array<size_t, 2>> pairs;
+    if (!one_atom_basisvectors_indices.empty()) {
+        pairs = one_atom_basisvectors_indices;
+    } else {
+        pairs.reserve(system1.getNumBasisvectors() * system2.getNumBasisvectors());
+        for (size_t col_1 = 0; col_1 < system1.getNumBasisvectors(); ++col_1) {
+            for (size_t col_2 = 0; col_2 < system2.getNumBasisvectors(); ++col_2) {
+                pairs.push_back({col_1, col_2});
             }
+        }
+    }
 
-            // Continue if the pair statet energy is not valid
-            double energy = std::real(system1.getHamiltonian().coeff(col_1, col_1) +
-                                      system2.getHamiltonian().coeff(col_2, col_2));
-            if (!checkIsEnergyValid(energy)) {
-                continue;
+    // Loop over the pairs
+    for (const auto &a : pairs) {
+        size_t col_1 = a[0];
+        size_t col_2 = a[1];
```

Comment:
Since we are using C++17 now, you can also use structured bindings for this type of loop:
```suggestion
    for (const auto [col_1, col_2] : pairs) {
```